### ISSUE TITLE
[NEW RULE] Minimize Deep Asynchronous Chains

### ISSUE THREAD

**Issue by ParasMindfire** [mindfiredigital/eslint-plugin-hub#29]
### 1️⃣ Is your rule proposal related to a problem? Please describe

Yes. When dealing with asynchronous operations, developers sometimes create very long chains of .then() calls for Promises or deeply nested await statements within a single async function. While functionally correct, these long sequences can significantly hinder code readability and maintainability. They increase cognitive load for developers trying to understand the flow, make debugging more difficult, and can obscure the distinct logical steps within the asynchronous process. This often leads to code that is harder to test and reuse.

### 2️⃣ Describe the solution you'd like

This rule should enforce a configurable limit on the "depth" of asynchronous operations within a single function scope.

For Promise chains, "depth" refers to the number of chained .then(), .catch(), or .finally() calls originating from a single Promise.

For async/await, "depth" refers to the number of await expressions within the top-level block of a single async function.

The rule should flag functions that exceed this configured maximum depth, encouraging developers to refactor the code into smaller, more manageable asynchronous functions or to use patterns like Promise.all for parallelizable operations.

```
// Depth 3 is okay
getUser()
  .then(user => processUser(user))         // 1
  .then(processed => saveUser(processed))   // 2
  .catch(error => handleError(error));      // 3
```

```
// Depth 4 - Triggers warning
fetch('/api/complex-data')
  .then(response => response.json())                // 1
  .then(data => firstProcessing(data))             // 2
  .then(intermediate => secondProcessing(intermediate)) // 3
  .then(finalData => displayResult(finalData));     // 4
```

```
// 5 awaits - Triggers warning
async function originalProblem(id) {
  const r1 = await op1(id);     // 1
  const r2 = await op2(r1);     // 2
  const r3 = await op3(r2);     // 3
  const r4 = await op4(r3);     // 4
  const r5 = await op5(r4);     // 5
  return r5;
}
```

```
// Refactored Solutions of chaining

function handleSecondHalf(intermediateData) {
  return secondProcessing(intermediateData)       // 1 (in this helper)
    .then(finalData => displayResult(finalData)); // 2 (in this helper)
}

fetch('/api/complex-data')
  .then(response => response.json())                // 1
  .then(data => firstProcessing(data))             // 2
  .then(intermediate => handleSecondHalf(intermediate)); // 3 (calls helper which returns a promise)
```

```
// Refactored Solutions of multiple awaits

async function handleSteps3to5(inputForStep3) {
  const r3 = await op3(inputForStep3); // 1 (in this helper)
  const r4 = await op4(r3);            // 2 (in this helper)
  const r5 = await op5(r4);            // 3 (in this helper)
  return r5;
}

async function refactoredSolution(id) {
  const r1 = await op1(id);        // 1
  const r2 = await op2(r1);        // 2
  return await handleSteps3to5(r2); // 3
}


```

### 3️⃣ Describe alternatives you've considered

Relying solely on code reviews: While helpful, manual detection is inconsistent and doesn't scale well. An automated lint rule provides immediate feedback.

Ignoring the issue: Leads to the problems of poor readability and maintainability described above.

Stricter functional programming approaches: While beneficial, they might be too prescriptive for all codebases. This rule offers a more direct and configurable check on a common imperative pattern.

### 4️⃣ Example configuration

Show how you expect the rule to be configured in an ESLint config file.

```json
{
  "rules": {
    "hub/minimize-deep-asynchronous-chains": ["warn", { "max": 3 }]
    "hub/no-excessive-async-chaining": ["warn", { "maxAwaitsPerFunction": 3, "maxPromiseThenChain": 3 }] // Alternative naming/options
  }
}
```

The max option would control the threshold for both promise chain links and await counts within a function.

Alternatively, separate options like maxAwaitsPerFunction and maxPromiseThenChain could be provided if finer-grained control is desired.

### 5️⃣ Additional context

Long asynchronous chains are a common source of "callback hell" like symptoms, even with Promises and async/await. This rule encourages developers to think about the structure of their asynchronous code, promoting:

Modularity: Breaking down complex async flows into smaller, named functions.

Readability: Making the primary execution path of a function easier to follow.

Testability: Smaller async units are easier to test in isolation.

Reusability: Extracted helper async functions can often be reused.

This rule would be particularly beneficial in codebases with heavy I/O operations or complex multi-step asynchronous workflows.
